src/piecewise.jl

```diff
@@ -0,0 +1,215 @@
+Domain() = Domain{:open, :open}(Inf, -Inf)
+
+lowerbound(x::Domain{L, R}) where {L, R} = (x.lo, L)
+upperbound(x::Domain{L, R}) where {L, R} = (x.hi, R)
+
+inf(x::Domain) = x.lo
+sup(x::Domain) = x.hi
+
+"""
+    rightof(val, lowerbound)
+
+Determine if a value is on the right of a lower bound.
+"""
+function rightof(x::Real, (val, bound))
+    bound == :closed && return val <= x
+    return val < x
+end
+
+function rightof((val1, bound1), (val2, bound2))
+    if val1 < val2 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+"""
+    leftof(val, upperbound)
+
+Determine if a value is on the left of an upper bound.
+"""
+function leftof(x::Real, (val, bound))
+    bound == :closed && return x <= val
+    return x < val
+end
+
+function leftof((val1, bound1), (val2, bound2))
+    if val2 < val1 || (val1 == val2 && bound1 == bound2 == :closed)
+        return false
+    end
+
+    return true
+end
+
+function leftof(d1::Domain, d2::Domain)
+    val1, bound1 = upperbound(d1)
+    val2, bound2 = lowerbound(d2)
+
+    val1 == val2 && return !(bound1 == bound2 == :closed)
+    return val1 < val2
+end
+
+Base.in(x::Real, domain::Domain) = rightof(x, lowerbound(domain)) && leftof(x, upperbound(domain))
+
+function Base.intersect(d1::Domain, d2::Domain)
+    left = max(lowerbound(d1), lowerbound(d2))
+    right = min(upperbound(d1), upperbound(d2))
+
+    left > right && return Domain()
+    return Domain(left, right)
+end
+
+function Base.isempty(domain::Domain)
+    lo, lobound = lowerbound(domain)
+    hi, hibound = upperbound(domain)
+
+    lo == hi && return !(lobound == hibound == :closed)
+    return lo > hi
+end
+
+
+struct Constant{T}
+    value::T
+end
+
+(constant::Constant)(::Any) = constant.value
+(constant::Constant)(::Interval) = interval(constant.value)
+
+# TODO Proper type parameters
+struct Piecewise
+    domain::Vector
+    fs
+    continuity::Vector{Int}
+    singularities::Vector
+end
+
+function Piecewise(
+        subdomains::Vector{<:Domain},
+        fs,
+        continuity::Vector{Int} = fill(-1, length(subdomains) - 1))
+
+    if length(subdomains) != length(fs)
+        throw(ArgumentError("the number of domains and the number of functions don't match"))
+    end
+    
+    if length(subdomains) - 1 != length(continuity)
+        n = length(subdomains)
+        throw(ArgumentError("$(length(sub)) junction points but $(n - 1) are expected based on the number of domains ($n)"))
+    end
+
+    for k in 1:length(subdomains) - 1
+        s1 = subdomains[k]
+        s2 = subdomains[k + 1]
+
+        if !leftof(s1, s2)
+           throw(ArgumentError("domains are either not ordered or not disjoint"))
+        end
+    end
+
+    singularities = sup.(subdomains[1:end-1])
+
+    return Piecewise(subdomains, fs, continuity, singularities)
+end
+
+function Piecewise(pairs::Vararg{Pair} ; continuity = fill(-1, length(pairs) - 1))
```

Comment:
By default `piecewise.continuity` is a vector containing `-1`, which means all points are discontinuous points (cf. `discontinuities` functions)?